Feature/gazebo-sim-fixes

MaxxWilson opened the pull request:
# PR Summary
PR Link: https://github.com/VEXU-GHOST/VEXU_GHOST/compare/develop...feature/gazebo-sim-fixes
Issue Link: https://github.com/VEXU-GHOST/VEXU_GHOST/issues/6

### Description
Develop a Plugin which mocks the V5SensorUpdate and V5ActuatorCommand ROS topics in Gazebo

### Reviewers
Tag reviewers.

- Required: @MaxxWilson
  
- Optional: @JakeWendling @xwilson03 

---
### Changelog
- Adds a diff drive gazebo model so that we could split testing the autonomy, controls, and estimation stack from developing the swerve model. 
- Adds EKF from robot_localization pkg. Somewhat tuned. Need to handle increase of estimated covariance over time.

### Reviewer Guide
Most changes are in the ghost_sim pkg. 

Field walls were removed since it caused a not-obvious bug: when given a cmd_vel, the robot collided with the walls and continued to attempt to achieve a cmd_vel, causing the wheels to slip, changing the odometry readings. Ultimately, the odometry encoder and the imu messages disagreed, causing the EKF to diverge. 

Note that the EKF in this PR uses 3 "sensors": gazebo IMU plugin, actual truth model data as "odometry" data, and gazebo diff drive plugin odometry configured as "WORLD" instead of "ENCODER". Therefore, the data given to the EKF is very ideal data. Performance should be monitored with more noisy data.

### Testing
#### Automatic
 - N/A
#### Manual
- launch the plotjuggler, diff drive, robot_localization stack using:
`ros2 launch ghost_sim test_tank_sim.launch.py`
in a separate terminal launch
`ros2 topic pub --once /cmd_vel geometry_msgs/msg/Twist "{linear: {x: 5.0, y: 0.0, z: 0.0}, angular: {x: 0.0, y: 0.0, z: 0.0}}"`
or 
`ros2 topic pub --once /cmd_vel geometry_msgs/msg/Twist "{linear: {x: 0.0, y: 0.0, z: 0.0}, angular: {x: 0.0, y: 0.0, z: 1.0}}"`
for sinsusoid motion. 
If subscribed to the odometry/filtered topic, the pose and twist should update with slight deviation from the gz/model_states pose and twist. 

### Documentation
- [Here]( https://docs.ros.org/en/melodic/api/robot_localization/html/index.html) is there package used for robot localization

### Checklist
- [x] Confirmed all tests pass on a clean build
- [X] Added reviewers in Github
- [x] Posted PR Summary to Discord PR's Channel
- [x] Ran uncrustify on any modified C++ files
- [x] Ran Colcon Lint for any modified CMakeLists.txt or Package.xml

[Here's]( https://github.com/VEXU-GHOST/VEXU_GHOST/assets/47638252/62d89e47-defd-467e-a05d-5d1a181e949a
) a video of the diff drive EKF

melissajcruz committed:
Compiles and runs sim with gz joint to encoder
sensor updates publishing across topic
Untested since gz motor command interface not defined

melissajcruz committed:
Does not compile
Adds encoder to joints and
motor to joint cmd interface

melissajcruz committed:
Merge branch 'develop' into feature/gazebo-sim-fixes
Updates gazebo sim fixes - up to date with new develop refactor

melissajcruz committed:
Merge branch 'develop' into feature/gazebo-sim-fixes

melissajcruz committed:
Sim runs. Robot control unstable b/c
of dummy v5actuator cmd data

melissajcruz committed:
cleans std::cout

melissajcruz committed:
v5actuatorcmd test pubs v5 mtr cmd
must double check units

melissajcruz committed:
sim rotates wheel module
but very slowly
double check units

melissajcruz committed:
Swerve moves in sim with pseudo turret. Must
validate with plots

melissajcruz committed:
Adds swerve kinematics and joystick msg

JakeWendling committed:
joystick publishes, sim torques dont work

melissajcruz committed:
joystick devel pt 2

MaxxWilson committed:
adds symlink-install option to build script

MaxxWilson committed:
modifes joystick channel to be configurable from launch file

MaxxWilson committed:
attempts to fix bugs

MaxxWilson committed:
uncomment swerve jacobian

JakeWendling committed:
Merge pull request #34 from VEXU-GHOST/bug-fix/sim-node-debugging

attempts to fix bugs

melissajcruz committed:
test diff drive model in gz.
no gz diff drive plugin loaded

melissajcruz committed:
Test diff drive spawns in gz
Able to send vel cmds
Need to validate vel/pos tracking

melissajcruz committed:
Adds plotjuggler to
test_tank launch file
for live streaming data

melissajcruz committed:
imu topic not being published to for robot_localization pkg

melissajcruz committed:
robot localization EKF node launches
validated with plotjuggler

melissajcruz committed:
robot localization node launches
need to validate filtered odom output
with plotjuggler

melissajcruz committed:
Merge branch 'feature/gazebo-sim-fixes' of github.com:VEXU-GHOST/VEXU_GHOST into feature/gazebo-sim-fixes

melissajcruz committed:
Merge pull request #38 from VEXU-GHOST/develop

Updates gazebo sim fixes branch with develop

melissajcruz committed:
EKF works with test diff drive if given cmd_vel. Need to test EKF when diff drive is stationary.
Must be tuned again for swerve drive.

melissajcruz commented:
This branch has V5Actuator update interface that lacks rigorous testing framework to validate.

In an attempt to overcome this, a test differential drive model was implemented without the V5actuator update interface using a differential drive gazebo plugin with the intention of combining the urdf model with the V5actuator update interface. However, development deviated from testing the V5actuator update interface to implementing the test diff drive model for developing an autonomy, estimation, control stack. 

So, this branch has a diff drive with a robot localization stack implemented via a somewhat tuned EKF. I think the main purpose of this branch has deviated from the original purpose and should be merged back into dev. 

I'm not sure what the status is of the sensor and actuator interfaces. We should discuss this!

melissajcruz committed:
Cleaning not needed files

melissajcruz committed:
Added dependencies for package,.xmls

MaxxWilson commented:
If wheels slipping is a problem, we might try increasing robot mass. On the real bot, drive motors should hit stall torque before wheels lose traction. Not really that big a deal, but could be nice to try.

MaxxWilson commented on .gitmodules:
```diff
@@ -24,3 +24,6 @@
 	path = 09_External/ThirdParty-Mumps
 	url = git@github.com:MaxxWilson/ThirdParty-Mumps.git
 	branch = with-src
+[submodule "09_External/PlotJuggler"]
```
We should also need to add plotjuggler_ros as a submodule: https://github.com/PlotJuggler/plotjuggler-ros-plugins/tree/rolling

IIRC, this is where you actually have to make source changes to add new msg types

melissajcruz replied:
I added plotjuggler_ros to .gitmodules file. 

Make source changes to make new msg types to stream to plotjuggler?

MaxxWilson replied:
These need to be added through git submodule add interface. Not just to this file.

MaxxWilson commented on 03_ROS/ghost_ros/CMakeLists.txt:
```diff
@@ -51,24 +51,25 @@ include_directories(include)
-install(TARGETS
-  foxglove_diagnostics_node
-  DESTINATION lib/${PROJECT_NAME})
+# add_executable(foxglove_diagnostics_node
```
Was planning to delete this FYI, bc Foxglove seems sorta unsupported. Will handle in separate PR.

melissajcruz replied:
It's been commented out in the CMake.

MaxxWilson commented on 03_ROS/ghost_sim/include/ghost_sim/test_publisher_v5_actuator_cmd.hpp:
```diff
@@ -33,6 +33,10 @@ class testPublisherV5ActuatorCmd : public rclcpp::Node {
 
 	void publishData();
 	rclcpp::Publisher<ghost_msgs::msg::V5ActuatorCommand>::SharedPtr output_pub_;
+	std::array<std::string, 8> motors_{"DRIVE_LEFT_FRONT_LEFT_MOTOR",
```
swap std::array for std::vector (which has dynamic sizing)?

melissajcruz replied:
This is acutally a test file that makes pseudo data as if you are sending motor cmds to the individual motors. This will not run on the robot in real time. So I think this implementation is fine for the scope of this file.

I can optimize this when the v5 actuator interface is finalized.

MaxxWilson commented on 03_ROS/ghost_sim/include/ghost_sim/v5_robot_plugin.hpp:
```diff
@@ -35,9 +36,13 @@ class V5RobotPlugin : public gazebo::ModelPlugin {
 	void jointToEncoderTransform();
-	ghost_msgs::msg::V5EncoderState wrapEncoderMsg(const int col_index);
-	ghost_msgs::msg::V5SensorUpdate populateSensorMsg();
+	Eigen::VectorXd motorToJointTransform(Eigen::VectorXd motor_data);
```
Header file methods need comment blocks. Autofills on my VSCode by doing /** and then enter

melissajcruz replied:
Changed.

MaxxWilson commented on 03_ROS/ghost_sim/include/ghost_sim/v5_robot_plugin.hpp:
```diff
@@ -35,9 +36,13 @@ class V5RobotPlugin : public gazebo::ModelPlugin {
-	ghost_msgs::msg::V5EncoderState wrapEncoderMsg(const int col_index);
-	ghost_msgs::msg::V5SensorUpdate populateSensorMsg();
+	Eigen::VectorXd motorToJointTransform(Eigen::VectorXd motor_data);
+	Eigen::VectorXd jointToMotorTransform(Eigen::VectorXd joint_data);
```
Should pass these by const ref if they aren't changing, and ref if they are changing. Try to avoid copying non-trivial types.

melissajcruz replied:
Changed to const ref.

MaxxWilson commented on 03_ROS/ghost_sim/src/v5_robot_plugin.cpp:
```diff
@@ -44,20 +45,40 @@ class V5RobotPluginPrivate {
+
+	// Note that the float32 torque_nm parameter of this msg type is ignored since
+	// torques can only be applied not measured in Gazebo
+	std::array<ghost_msgs::msg::V5EncoderState, 21> encoder_msg_;
```
switch std::array to std::vector, array is fixed compile time size, idk why we would ever use it over vector.

melissajcruz replied:
Will be changed when v5 actuator interfaces are finalized.

MaxxWilson commented on 03_ROS/ghost_sim/src/v5_robot_plugin.cpp:
```diff
@@ -71,6 +92,15 @@ class V5RobotPluginPrivate {
+	ghost_msgs::msg::V5Joystick joystick_msg_;
+
+	// Constants
+	float DEG_TO_RAD = M_PI / 180;
```
can these be 1) either const or constexpr, and 2) moved to ghost_util into a "types.hpp" header. Might want to use somewhere else.

melissajcruz replied:
This is already defined in 01_Libraries/ghost_common/include/ghost_common/util/angle_util.hpp so I just replaced the constants with the constants that already exist.

MaxxWilson commented on 03_ROS/ghost_sim/src/v5_robot_plugin.cpp:
```diff
@@ -193,68 +257,153 @@ void V5RobotPlugin::Load(gazebo::physics::ModelPtr model, sdf::ElementPtr sdf){
+void V5RobotPlugin::wrapEncoderMsg(){
 	int col_index = 0;
-	// Lamda for-each expression to get encoder values for every joint
-	for_each(begin(impl_->joint_msg_list_), end(impl_->joint_msg_list_), [&](const std::string &joint_data){
```
might be cleaner / more consistent with rest of codebase to use:

for(const std::string& name : impl_->joint_msg_list_){

or

for(const auto& name : impl_->joint_msg_list_){

melissajcruz replied:
Are lambda expressions avoided in this repo?

MaxxWilson commented on 03_ROS/ghost_sim/src/v5_robot_plugin.cpp:
```diff
@@ -193,68 +257,153 @@ void V5RobotPlugin::Load(gazebo::physics::ModelPtr model, sdf::ElementPtr sdf){
-	// Lamda for-each expression to get encoder values for every joint
-	for_each(begin(impl_->joint_msg_list_), end(impl_->joint_msg_list_), [&](const std::string &joint_data){
-			// Get iterator that points to corresponding joint name in joint_names_
-			auto joint_name_itr = find(impl_->joint_names_.begin(), impl_->joint_names_.end(), joint_data);
```
is this something that could be more efficient if we had an unordered_map for looking this up from name?

melissajcruz replied:
These two lines were deleted and do not exist in my branch locally?

But unordered_map would be better for future applications.

MaxxWilson commented on 03_ROS/ghost_sim/src/v5_robot_plugin.cpp:
```diff
@@ -193,68 +257,153 @@ void V5RobotPlugin::Load(gazebo::physics::ModelPtr model, sdf::ElementPtr sdf){
 }
 
+// Preserves order of joints listed in xacro
 void V5RobotPlugin::getJointStates(){
```
could this be updateJointStates(), since it isn't returning anything?

melissajcruz replied:
Changed!

MaxxWilson commented:
Looks very solid! And we have diff drive actually working in sim, which is dope.

There may be a lot of small interface changes on upcoming robot_interface overhaul PR, but those will be on my end lol.

Is there a way to add unit testing so we know we can load the plugin and it can load a default config file? I think there is a GazeboTestFixture example somewhere.

melissajcruz committed:
Adds changes from PR comments

melissajcruz committed:
Addresses PR comments and adds gz tests
testing is in development

melissajcruz committed:
update git ignore

melissajcruz committed:
Testing script runs, fails some assertions.

melissajcruz committed:
Fixed weird x angular velocty:
Make sure base model and diff drive plugin params agree!

melissajcruz commented:
Final Update:

After speaking with Maxx, determined Gazebo Test Fixture used seems to be doing some dirty deallocation at the end of the test scripts. So, if your output looks something like this:

```~/VEXU_GHOST$ ./scripts/test_diff_drive.sh
xacro: in-order processing became default in ROS Melodic. You can drop the option.
Allow generated test description files to be moved to temp testing folder
Killing gazebo processes
gzserver: no process found
UpdateCTestConfiguration  from :/home/melcruz/VEXU_GHOST/build/ghost_sim/CTestConfiguration.ini
Parse Config file:/home/melcruz/VEXU_GHOST/build/ghost_sim/CTestConfiguration.ini
 Add coverage exclude regular expressions.
UpdateCTestConfiguration  from :/home/melcruz/VEXU_GHOST/build/ghost_sim/CTestConfiguration.ini
Parse Config file:/home/melcruz/VEXU_GHOST/build/ghost_sim/CTestConfiguration.ini
Test project /home/melcruz/VEXU_GHOST/build/ghost_sim
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 1
    Start 1: test_diff_drive

1: Test command: /usr/bin/python3.10 "-u" "/opt/ros/humble/share/ament_cmake_test/cmake/run_test.py" "/home/melcruz/VEXU_GHOST/build/ghost_sim/test_results/ghost_sim/test_diff_drive.gtest.xml" "--package-name" "ghost_sim" "--output-file" "/home/melcruz/VEXU_GHOST/build/ghost_sim/ament_cmake_gtest/test_diff_drive.txt" "--command" "/home/melcruz/VEXU_GHOST/build/ghost_sim/test_diff_drive" "--gtest_output=xml:/home/melcruz/VEXU_GHOST/build/ghost_sim/test_results/ghost_sim/test_diff_drive.gtest.xml"
1: Test timeout computed to be: 120
1: -- run_test.py: invoking following command in '/home/melcruz/VEXU_GHOST/build/ghost_sim':
1:  - /home/melcruz/VEXU_GHOST/build/ghost_sim/test_diff_drive --gtest_output=xml:/home/melcruz/VEXU_GHOST/build/ghost_sim/test_results/ghost_sim/test_diff_drive.gtest.xml
1: [==========] Running 1 test from 1 test suite.
1: [----------] Global test environment set-up.
1: [----------] 1 test from GazeboRosDiffDriveTest
1: [ RUN      ] GazeboRosDiffDriveTest.testBringUp
1: [Msg] Waiting for master.
1: [Msg] Connected to gazebo master @ http://127.0.0.1:11345
1: [Msg] Publicized address: 172.25.171.127
1: [Msg] Loading world file [/home/melcruz/VEXU_GHOST/install/ghost_sim/share/ghost_sim/urdf/spin_up.world]
1: [Wrn] [Event.cc:61] Warning: Deleting a connection right after creation. Make sure to save the ConnectionPtr from a Connect call
1: [Dbg] [ServerFixture.cc:211] ServerFixture load in 2.9 seconds, timeout after 600 seconds
1: /home/melcruz/VEXU_GHOST/install/ghost_sim/share/ghost_sim
1: ===Loading diff drive model===
1: [INFO] [1706057065.861770653] [gazebo_ros_state]: Publishing states of gazebo models at [/model_states]
1: [INFO] [1706057065.862330012] [gazebo_ros_state]: Publishing states of gazebo links at [/link_states]
1: [INFO] [1706057066.114238301] [gazebo_ros_diff_drive]: Wheel pair 1 separation set to [0.300000m]
1: [INFO] [1706057066.114302665] [gazebo_ros_diff_drive]: Wheel pair 1 diameter set to [0.200000m]
1: [INFO] [1706057066.115204849] [gazebo_ros_diff_drive]: Subscribed to [/cmd_vel]
1: [INFO] [1706057066.116717739] [gazebo_ros_diff_drive]: Advertise odometry on [/odom]
1: [INFO] [1706057066.120285924] [gazebo_ros_diff_drive]: Publishing odom transforms between [odom] and [base_link]
1: [INFO] [1706057066.120324185] [gazebo_ros_diff_drive]: Publishing wheel transforms between [base_link], [left_wheel_hinge] and [right_wheel_hinge]
1: [Dbg] [ServerFixture.cc:130] ServerFixture::Unload
1: -- run_test.py: return code -11
1: -- run_test.py: generate result file '/home/melcruz/VEXU_GHOST/build/ghost_sim/test_results/ghost_sim/test_diff_drive.gtest.xml' with failed test
1: -- run_test.py: verify result file '/home/melcruz/VEXU_GHOST/build/ghost_sim/test_results/ghost_sim/test_diff_drive.gtest.xml'
1/1 Test #1: test_diff_drive ..................***Failed    5.30 sec

0% tests passed, 1 tests failed out of 1

Label Time Summary:
gtest    =   5.30 sec*proc (1 test)

Total Test time (real) =   5.31 sec

The following tests FAILED:
          1 - test_diff_drive (Failed)
Errors while running CTest
Output from these tests are in: /home/melcruz/VEXU_GHOST/build/ghost_sim/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.```

The diff drive test passed! If someone has a fix for this please let me know. 
Sorry Xander.

melissajcruz committed:
Changes diff drive tests and adds test script
The tests have some bad deallocation so you should visually test the diff drive in GZ
instead of running this test script.

melissajcruz committed:
Merge branch 'develop' of github.com:VEXU-GHOST/VEXU_GHOST into feature/gazebo-sim-fixes

melissajcruz committed:
Refactors sim stack after hardware refactor

melissajcruz committed:
Merge branch 'develop' into feature/gazebo-sim-fixes

melissajcruz committed:
Cleaning

MaxxWilson commented:
2/1 Test Failures:
![image](https://github.com/VEXU-GHOST/VEXU_GHOST/assets/47650195/a57c272f-52df-467a-b4fe-1fc543357b9f)

MaxxWilson committed:
moves world files to world folder

MaxxWilson committed:
fixes last world file move ref

MaxxWilson committed:
change world path to worlds

MaxxWilson committed:
moves test files to /tmp/
small cleanup
passing tests, approved!

MaxxWilson committed:
removes boilerplate testing from ghost_tank

MaxxWilson commented on 01_Libraries/ghost_control/include/ghost_control/models/dc_motor_model.hpp:
```diff
@@ -66,6 +66,13 @@ class DCMotorModel {
+	 *
+	 * @param voltage_mv voltage in millivolts
+	 */
+	void setMotorEffortMillivolts(double voltage_mv);
```
Where did these get added? Should be removed if they dont match develop?

MaxxWilson commented on 01_Libraries/ghost_control/include/ghost_control/motor_controller.hpp:
```diff
@@ -53,6 +53,24 @@ class MotorController {
+	float getPositionCommand(){
+		return des_pos_encoder_;
+	}
+
```
Where did these get added? Should be removed if they dont match develop?

MaxxWilson commented on 03_ROS/deprecated/ghost_ros/CMakeLists.txt:
```diff
@@ -51,24 +51,25 @@ include_directories(include)
 #####################################
 ##### Foxglove Diagnostics Node #####
 #####################################
```
Not sure why these didn't sync from develop, but lets remove entirely. They are deprecated.

MaxxWilson committed:
test fixing merge issues

MaxxWilson committed:
more reverting merge changes

MaxxWilson committed:
merge fixes v3

MaxxWilson committed:
final merge fixes?

MaxxWilson committed:
revert swerve changes

MaxxWilson reviewed (commented):
Good with this once submodules are fixed

melissajcruz committed:
Changes dep to add pj and pjros

melissajcruz committed:
Merge branch 'feature/gazebo-sim-fixes' of github.com:VEXU-GHOST/VEXU_GHOST into feature/gazebo-sim-fixes

melissajcruz committed:
Cleaning

melissajcruz reviewed (approved):
lgtm!
